docs/proposals/role_decomposition.md

```diff
@@ -0,0 +1,322 @@
+# Scaffolding for decomposing large roles
+
+## Why?
+
+Currently we have roles that are very large and encompass a lot of different
+components. This makes for a lot of logic required within the role, can
+create complex conditionals, and increases the learning curve for the role.
+
+## How?
+
+Creating a guide on how to approach breaking up a large role into smaller,
+component based, roles. Also describe how to develop new roles, to avoid creating
+large roles.
+
+## Proposal
+
+Create a new guide or append to the current contributing guide a process for
+identifying large roles that can be split up, and how to compose smaller roles
+going forward.
+
+### Large roles
+
+A role should be considered for decomposition if it:
+
+1) Configures/installs more than one product.
+1) Can configure multiple variations of the same product that can live
+side by side.
+1) Has different entry points for upgrading and installing a product
+
+Large roles should be responsible for:
+
+1) Composing smaller roles to provide a full solution such as an Openshift Master
```

Comment:
This is part of what I was hoping to outline with the Small roles section. I agree, we would need to have a defined method of splitting up and composing roles so that we do not have scenarios where, as Jan suggests, we are overly verbose.

My thought is we would have to identify components that make up the larger role and split those out. So if we had a role for `etcd` for example, that role would be responsible for doing everything that was required to get an `etcd` instance stood up -- creating its certificates, ensuring the correct version was installed, performing any upgrades that may be necessary if moving to a newer version, etc.

I would like to get some more feedback/input on how we would identify these components and how we would avoid being overly verbose. Perhaps we would want to have a role for example that contained all commonly used handlers so we could avoid copy/pasta for handlers such as restarting masters (this is used in a couple different places). It could then be clearly defined as a dependency in meta/main.yml